Closes #7069: 3.18 - Exclusions part

Miraeld opened the pull request:
# Description

Fixes #7069 
It allows the user to use a new filter `rocket_exclude_locally_host_fonts` to exclude some fonts from the feature.

## Type of change

- [x] New feature (non-breaking change which adds functionality).

## Detailed scenario

Just add some fonts to exclusiosn through the new filter (`rocket_exclude_locally_host_fonts`), and it won't be downloaded.

## Technical description

### Documentation

This pull request introduces a new feature to exclude specific fonts from being rewritten in the media fonts optimization process. The changes involve adding methods to handle font exclusions and updating relevant tests to ensure the new functionality works as expected.

### New Feature: Font Exclusions

* [`inc/Engine/Media/Fonts/Frontend/Controller.php`](diffhunk://#diff-3f1e418966792d4b51109c090fb00e9fe1ec75469559e0ce738d27d386e0f974R83-R95): Added methods `get_exclusions` and `is_excluded` to obtain and check font exclusion patterns, and updated `rewrite_fonts` to skip excluded fonts. [[1]](diffhunk://#diff-3f1e418966792d4b51109c090fb00e9fe1ec75469559e0ce738d27d386e0f974R83-R95) [[2]](diffhunk://#diff-3f1e418966792d4b51109c090fb00e9fe1ec75469559e0ce738d27d386e0f974R218-R253)
* [`inc/Engine/Optimization/DynamicLists/DynamicLists.php`](diffhunk://#diff-6b68fcd8111874aceb9796f87da85abbc6d9f16e186d1d6fa7b4354f1af41b98R334-R344): Added method `get_exclude_media_fonts` to retrieve excluded font templates.
* [`inc/Engine/Optimization/DynamicLists/Subscriber.php`](diffhunk://#diff-d900b5653489b76be1402afa7965c01db23cae07fb02b280bee69e7ff988b4efR48): Subscribed to the `rocket_media_fonts_exclusions` event and added `add_media_fonts_exclusions` method to include media font exclusions. [[1]](diffhunk://#diff-d900b5653489b76be1402afa7965c01db23cae07fb02b280bee69e7ff988b4efR48) [[2]](diffhunk://#diff-d900b5653489b76be1402afa7965c01db23cae07fb02b280bee69e7ff988b4efR216-R226)

### Tests and Fixtures Updates

* [`tests/Fixtures/inc/Engine/Media/Fonts/Frontend/Subscriber/HTML/expected_v1_excluded_v2.php`](diffhunk://#diff-c35d52875f48d81af3fd2ed5d712f12e41310750ef2b901a3dd834c31c24e2f7R1-R41): Added a new fixture for testing the exclusion of V2 fonts while rewriting V1 fonts.
* [`tests/Fixtures/inc/Engine/Media/Fonts/Frontend/Subscriber/rewriteFonts.php`](diffhunk://#diff-be1c59942b94721673370154f37badf0ba45cb89b2fa4e9476f5774c9bcd2eecR75-R97): Added a new test case `testShouldRewriteV1AndExcludeV2Fonts` to verify the exclusion functionality.
* [`tests/Integration/inc/Engine/Media/Fonts/Frontend/Subscriber/rewriteFonts.php`](diffhunk://#diff-f8cb3ab9e3c8a35724dfde51d694d01238dca1e489d99ef7a45e06d0b0eeefb3R26-R35): Updated setup and teardown methods to include the new `rocket_exclude_locally_host_fonts` filter, and added the `exclude_locally_host_fonts` method. [[1]](diffhunk://#diff-f8cb3ab9e3c8a35724dfde51d694d01238dca1e489d99ef7a45e06d0b0eeefb3R26-R35) [[2]](diffhunk://#diff-f8cb3ab9e3c8a35724dfde51d694d01238dca1e489d99ef7a45e06d0b0eeefb3R62-R65)

### New dependencies

None

### Risks

None

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

Khadreal committed:
:close #7069 Add exclusion for media font

Miraeld committed:
Merge branch 'feature/host-google-fonts' into feature/7069-exclusion-part

Miraeld committed:
Add exclusion feature + linter

Miraeld committed:
Add test

Miraeld committed:
Fix phpstan

Miraeld committed:
Fix type of returned value

Miraeld committed:
Fix unit tests

remyperona commented on inc/Engine/Media/Fonts/Frontend/Controller.php:
```diff
@@ -207,6 +215,42 @@ public function disable_google_fonts_preload( $disable ): bool {
+	 */
+	private function is_excluded( string $url, array $exclusions ): bool {
+		if ( ! empty( $exclusions ) ) {
+			foreach ( $exclusions as $exclusion ) {
+				if ( preg_match( '/' . $exclusion . '/', $url ) ) {
```
aren't we going to have an issue if the excluded pattern contains a `/` character? We probably needs some escaping

remyperona replied:
I don't know if one solution is better than the other, but to avoid the loop we previously did the following:
`$exclusions_str = implode( '|', $exclusions );`

remyperona commented on inc/Engine/Media/Fonts/Frontend/Controller.php:
```diff
@@ -207,6 +215,42 @@ public function disable_google_fonts_preload( $disable ): bool {
+	 * @return bool True if the URL is excluded, false otherwise.
+	 */
+	private function is_excluded( string $url, array $exclusions ): bool {
+		if ( ! empty( $exclusions ) ) {
```
maybe we can bail-out early if it's empty

jeawhanlee commented:
I might be mistaken but I'm not sure this PR considers this thread [here](https://github.com/wp-media/wp-rocket/issues/7069#issuecomment-2463021396) and the proposals following because as it is we might not be correctly excluding combined fonts or not?

Miraeld committed:
Add exclusion to GF combine + tests

Miraeld committed:
Comply feedbacks + add new tests

Miraeld committed:
Apply linter

Miraeld committed:
Merge branch 'feature/host-google-fonts' into feature/7069-exclusion-part

Miraeld committed:
Fix linter

Miraeld committed:
Fix Unit tests

Miraeld commented:
@remyperona @jeawhanlee 
Hello hello, 

I've made modifications to comply with feedbacks. 
Added, also, the exclusion to the Google Font combine feature so we won't combine excluded fonts. 

------

## Summary of changes by this PR:
This pull request introduces several enhancements to the font optimization process, specifically focusing on excluding certain fonts from being optimized based on user-defined patterns. These changes include updates to the core functionality, new methods for exclusion handling, and corresponding test cases.

### Core Functionality Enhancements:

* [`inc/Engine/Media/Fonts/Frontend/Controller.php`](diffhunk://#diff-3f1e418966792d4b51109c090fb00e9fe1ec75469559e0ce738d27d386e0f974R86-R103): Added methods `get_exclusions` and `is_excluded` to retrieve exclusion patterns and check if a font URL matches any exclusion pattern. Updated `rewrite_fonts` method to skip excluded fonts during font replacement. [[1]](diffhunk://#diff-3f1e418966792d4b51109c090fb00e9fe1ec75469559e0ce738d27d386e0f974R86-R103) [[2]](diffhunk://#diff-3f1e418966792d4b51109c090fb00e9fe1ec75469559e0ce738d27d386e0f974R255-R298)

### Integration with Dynamic Lists:

* [`inc/Engine/Optimization/DynamicLists/DynamicLists.php`](diffhunk://#diff-6b68fcd8111874aceb9796f87da85abbc6d9f16e186d1d6fa7b4354f1af41b98R334-R344): Introduced `get_exclude_media_fonts` method to fetch excluded font templates.
* [`inc/Engine/Optimization/DynamicLists/Subscriber.php`](diffhunk://#diff-d900b5653489b76be1402afa7965c01db23cae07fb02b280bee69e7ff988b4efR48): Added event subscription for `rocket_media_fonts_exclusions` and implemented `add_media_fonts_exclusions` to merge dynamic list exclusions with existing exclusions. [[1]](diffhunk://#diff-d900b5653489b76be1402afa7965c01db23cae07fb02b280bee69e7ff988b4efR48) [[2]](diffhunk://#diff-d900b5653489b76be1402afa7965c01db23cae07fb02b280bee69e7ff988b4efR216-R226)

### Google Fonts Optimization:

* [`inc/Engine/Optimization/GoogleFonts/AbstractGFOptimization.php`](diffhunk://#diff-55886ca90f690e13573f1b2dd81c6d92bc8cfc10ae7cb971f96c40f38dfaebe3R133-R181): Added methods `get_exclusions` and `is_excluded` to handle exclusion patterns for Google Fonts optimization.
* `inc/Engine/Optimization/GoogleFonts/Combine.php` and `CombineV2.php`: Updated `optimize` method to filter out excluded fonts before processing. [[1]](diffhunk://#diff-8a071302d5007e4f981805795db143f7a2511decb61ae685a46fa7a5ab16d8f3L60-R79) [[2]](diffhunk://#diff-8d92fa3e24b9097c3851c553655b9f6381313d89e3ccd9fec2260bf01ff5c4d9L43-R63)

### Test Cases:

* Added new test fixtures and updated existing ones to validate the exclusion functionality. [[1]](diffhunk://#diff-c35d52875f48d81af3fd2ed5d712f12e41310750ef2b901a3dd834c31c24e2f7R1-R41) [[2]](diffhunk://#diff-7191a0cddf064195e55b5f8ae204a692fd04e457fd47072aa37bd717bf45969aR1-R12) [[3]](diffhunk://#diff-7f29f7bd249ab98ca19a8834021a9d21b7265d97226a631a1ca45ee4f71a9e93R1-R12) [[4]](diffhunk://#diff-be1c59942b94721673370154f37badf0ba45cb89b2fa4e9476f5774c9bcd2eecR75-R97) [[5]](diffhunk://#diff-be1c59942b94721673370154f37badf0ba45cb89b2fa4e9476f5774c9bcd2eecR117-R136) [[6]](diffhunk://#diff-f2185c7d46f0c602fef9c4aa01af3f639ae8fcfa912f5e4bd998dc807246bde7R332-R370) [[7]](diffhunk://#diff-4f13b58024ddd5f364f28d001989eab7ab584e331d309fd5726802f708a78cc8R226-R264)
* Updated integration tests to include the new exclusion filters and ensure they are applied correctly. [[1]](diffhunk://#diff-f8cb3ab9e3c8a35724dfde51d694d01238dca1e489d99ef7a45e06d0b0eeefb3R26-R35) [[2]](diffhunk://#diff-f8cb3ab9e3c8a35724dfde51d694d01238dca1e489d99ef7a45e06d0b0eeefb3R62-R65) [[3]](diffhunk://#diff-6f9eb0be96cb0a8c61b98ca609ee5f9c5263bc0e4a094fd10237af6ebaee6b8eR19-R20) [[4]](diffhunk://#diff-6f9eb0be96cb0a8c61b98ca609ee5f9c5263bc0e4a094fd10237af6ebaee6b8eR33-R41) [[5]](diffhunk://#diff-6f9eb0be96cb0a8c61b98ca609ee5f9c5263bc0e4a094fd10237af6ebaee6b8eR52-R68) [[6]](diffhunk://#diff-6f9eb0be96cb0a8c61b98ca609ee5f9c5263bc0e4a094fd10237af6ebaee6b8eR111-R114)

Miraeld committed:
Fix integration tests after merge

hanna-meda commented:
@Miraeld, thank you for this PR. 
I could only make the filter work using font family name, i.e. `$exclusions[] = 'Roboto';` 
Based on this [AC](https://github.com/wp-media/wp-rocket/issues/7069#issue-2616816387) _The exclusion should be regex compatible_, we should also cover expressions such as:
`$exclusions[] = 'https://fonts.googleapis.com/css?family=.*';`
`$exclusions[] = 'fonts.googleapis.com/.*';`
`$exclusions[] = 'https://fonts.googleapis.com/css?family=Noto%20Sans%3A400%2C700%7CMontserrat%3A700&display=swap';`
`$exclusions[] = 'https://fonts.googleapis.com/css?family=Rob.*';`

Returning this back to In Progress based on this comment. cc @piotrbak

remyperona commented on inc/Engine/Media/Fonts/Frontend/Controller.php:
```diff
@@ -244,6 +252,50 @@ public function disable_google_fonts_preload( $disable ): bool {
+		// Escape each exclusion pattern to prevent regex issues.
+		$escaped_exclusions = array_map(
+				function ( $exclusion ) {
+					return preg_quote( $exclusion, '#' ); // '#' is used as the delimiter.
```
Thinking back on this, preg_quote() is going to prevent RegEx characters from doing what we need (as reported by Hanna), so the escaping should only be done for the `#` character, using a `str_replace( '#', '\#', string )` for example

piotrbak commented:
@hanna-meda after discussion on the daily, since the URL doesn't matter for the exclusions, we'll match the regex after it:
- `family(.*)Montserrat%3A700&display=swap`
-`family=Rob(.*)o`

cc @jeawhanlee

remyperona committed:
Merge branch 'feature/host-google-fonts' into feature/7069-exclusion-part

remyperona committed:
update formatting for exclusions

remyperona committed:
add additional escaping

jeawhanlee committed:
Updated test

jeawhanlee committed:
Updated sanitization

jeawhanlee committed:
Use trait in classes

jeawhanlee committed:
Merge branch 'feature/host-google-fonts' into feature/7069-exclusion-part

Mai-Saad commented:
Thanks for the update.
[WIP] exploratory test notes:

1. If the HTML have already combined file , the feature won't be applied ex this page  (no font is downloaded nor rewritten) https://new.rocketlabsqa.ovh/combinedalready
2. Exclusion is now case sensitive i.e` Montserrat `will work while `montserrat` won't (test page https://new.rocketlabsqa.ovh/combine_gf_v2new)
3. Using exclusion like this `$exclusions[] = 'family=Oswald&#038';` won't make the feature work at all (nothing is downloaded nor rewritten, only combined external)  here https://new.rocketlabsqa.ovh/combine_gf_v2new
4. Exclusion by argument is not working i.e `$exclusions[]= 'testRobo';` here https://new.rocketlabsqa.ovh/combine_gf_v2new
5. If the filter returns `[""]`, the feature won't work at all (fonts will be there , not combined nor locally hosted) and nothing in debug.log about that

@piotrbak What do you think is a must to be fixed here? or other GH

Note: these exclusions are working fine here https://new.rocketlabsqa.ovh/combine_gf_v2new (partial match of the font URL, partial match with regex)
 ```
$exclusions[] = 'Lora'; 
$exclusions[] = 'family=Phetsarath';
$exclusions[] = 'family=R(.*)';
$exclusions[] = 'family=Phetsarath:wght@400;700&display=swap';

```

piotrbak commented:
@Mai-Saad 
1. Seems to be a markup issue, lack of `"`
2. That's fine

Points 3 and 4 seem to be related, is it possible that we're trying to exclude **after** the combination process 🤔 ?

5. It's fine

Mai-Saad commented:
For 3 `family=Phetsarath:wght@400;700&display=swap` is working {excluding only related font} but `family=Phetsarath:wght@400;700&#038;display=swap` is not working  (the one not working is from the combined file in HTML , while the working one is when we open combined in browser)

remyperona commented:
Point 4 is not working because the exclusion is done against the URL only, not attributes of the link tag.

remyperona committed:
Merge branch 'feature/host-google-fonts' into feature/7069-exclusion-part

remyperona committed:
make pattern case insensitive

remyperona commented:
I fixed point 2 with the latest commit.

Point 3 happens because of the difference between `&` and `&#038;`.

The combine GF and self-host use the same exclusion filter, but the format of the URL is different between them: combine gf uses `&` and then transform the URL to use `&#038;`

So if you use `&#038;`, the exclusion won't work for combine GF, but will work for self-host, excluding the full combined URL.
